serverless/cmd/feeder/github/main.go

```diff
@@ -0,0 +1,536 @@
+	/*
+	   [branch "master"]
+	   	remote = upstream
+	   	merge = refs/heads/master
+	*/
+	return forkRepo, nil
+}
+
+func authWithGithub(ctx context.Context, ghToken string) (*github.Client, error) {
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: ghToken})
+	tc := oauth2.NewClient(ctx, ts)
+	return github.NewClient(tc), nil
+}
+
+func pullAndGetRepoHead(r *git.Repository) (*plumbing.Reference, *git.Worktree, error) {
+	// Figure out where the worktree is for repo 'r'.
+	wt, err := r.Worktree()
+	if err != nil {
+		return nil, nil, fmt.Errorf("workTree(%v) err: %v", r, err)
+	}
+	// Pull the latest commits from remote 'origin'.
+	if err := wt.Pull(&git.PullOptions{RemoteName: "origin"}); err != nil && err != git.NoErrAlreadyUpToDate {
+		return nil, nil, fmt.Errorf("git pull %v/origin err: %v", r, err)
+	}
+	// Get the master HEAD - we'll need this to branch from later on.
+	headRef, err := r.Head()
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading %v HEAD ref err: %v", r, err)
+	}
+	// Print the latest commit that was just pulled
+	commit, err := r.CommitObject(headRef.Hash())
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading %v HEAD commit err: %v", r, err)
+	}
+	glog.V(1).Infof("Reading %v: HEAD commit:\n%v", r, commit)
+	return headRef, wt, nil
+}
+
+func createCheckpointPR(ctx context.Context, opts *options, forkRepo *git.Repository, ghCli *github.Client) error {
+	// Pull forkRepo and get the ref for origin/master branch HEAD.
+	headRef, workTree, err := pullAndGetRepoHead(forkRepo)
+	if err != nil {
+		return err
+	}
+
+	// TODO: if checkpoint and checkpoint.witnessed bodies are different (strictly, if checkpoint is newer),
+	// then we need to be feeding checkpoint to the witness, otherwise we can feed the witnessed one and
+	// short-circuit creating a PR if our witness(es) has/have already signed it.
+	inputCP := fmt.Sprintf("%s/checkpoint", filepath.Join(opts.witnessClonePath, opts.logRepoPath))
+	glog.V(1).Infof("Reading CP from %q", inputCP)
+	cp, err := ioutil.ReadFile(inputCP)
+	if err != nil {
+		return fmt.Errorf("failed to read input checkpoint: %v", err)
+	}
+
+	// 2. kick off feeder.
+	glog.V(1).Infof("CP to feed:\n%s", string(cp))
+	wCp, err := feed(ctx, cp, opts)
+	if err != nil {
+		return err
+	}
+
+	witnessNote := string(wCp)
+	lines := strings.Split(witnessNote, "\n")
+	if len(lines) < 3 {
+		return fmt.Errorf("witnessed checkpoint Note seems to have too few lines: \n%v", witnessNote)
+	}
+	// TODO: can we use actual note struct to do this.
+	cpSize, cpHash := lines[1], lines[2]
+
+	glog.V(1).Infof("CP after feeding:\n%s", witnessNote)
+
+	if bytes.Equal(cp, wCp) {
+		fmt.Println("No signatures added")
+		return nil
+	}
+
+	// Create a git branch for the witnessed checkpoint to be added, named
+	// using the base64(checkpoint hash).  Construct a fully-specified
+	// reference name for the new branch, then create using git plumbing.
```

Comment:
'create using git plumbing' is out of date.

Reply:
I've removed that bit of the comment for now.